src/charm.py

```diff
@@ -431,6 +439,103 @@ def _on_new_vault_kv_client_attached(self, event: NewVaultKvClientAttachedEvent)
+        """
+        if not self._is_relation_created(S3_RELATION_NAME):
+            event.fail(message="S3 relation not created. Failed to perform backup operation.")
+            return
+
+        if not self.unit.is_leader():
+            event.fail(message="Only leader unit can perform backup operations.")
+            return
+
+        missing_parameters = self._get_missing_s3_parameters()
+        if missing_parameters:
+            event.fail(message=f"S3 parameters missing. {missing_parameters}")
+            return
+
+        s3_parameters = self._retrieve_s3_parameters()
+
+        if not (session := self._create_s3_session(s3_parameters)):
+            event.fail(message="Failed to create S3 session.")
+            return
+
+        try:
+            self._create_s3_bucket(
+                session=session,
+                region=s3_parameters["region"],
+                bucket_name=s3_parameters["bucket"],
+                endpoint=s3_parameters["endpoint"],
+            )
+            logger.info("Created S3 bucket %s", s3_parameters["bucket"])
+        except KeyError as e:
+            logger.error("Missing required S3 parameter: %s", e)
+            event.fail(message="Failed to create S3 bucket.")
+            return
+        except (BotoCoreError, ClientError) as e:
+            logger.error("Failed to create S3 bucket: %s", e)
+            event.fail(message="Failed to create S3 bucket.")
+            return
+        snapshot = self._create_raft_snapshot()
+        if not snapshot:
+            event.fail(message="Failed to create raft snapshot.")
+            return
+        backup_key = self._upload_content_to_s3(
+            session=session,
+            content=snapshot,
+            bucket_name=s3_parameters["bucket"],
+            endpoint=s3_parameters["endpoint"],
+        )
+        if not backup_key:
+            event.fail(message="Failed to upload backup to S3 bucket.")
+            return
+        logger.info("Backup uploaded to S3 bucket %s", s3_parameters["bucket"])
+        event.set_results({"backup-id": backup_key})
+
+    def _upload_content_to_s3(
+        self,
+        session: boto3.session.Session,
+        content: requests.Response,
+        bucket_name: str,
+        endpoint: str,
+    ) -> Optional[str]:
+        """Uploads the provided contents to the provided S3 bucket.
+
+        Args:
+            session: S3 session.
+            content: Byte contents to upload.
+            bucket_name: S3 bucket name.
+            endpoint: S3 endpoint.
+
+        Returns:
+            str: The S3 key of the uploaded content.
+        """
+        key = self._get_backup_key()
+        try:
+            s3 = session.resource("s3", endpoint_url=endpoint)
+            bucket = s3.Bucket(bucket_name)
+            bucket.put_object(Key=key, Body=content)
+        except (BotoCoreError, ClientError) as e:
+            logger.warning("Error uploading content to bucket %s: %s", bucket_name, e)
```

Comment:
Found another one that might be better at the `error` level.

Reply:
✅